core/src/main/java/org/apache/druid/segment/column/Types.java

```diff
@@ -112,4 +121,529 @@
+   * total byte size of serializing the array is larger than the max size parameter, this method will explode via a
+   * call to {@link #checkMaxBytes}.
+   *
+   * layout: | null (byte) | size (int) | {| null (byte) | size (int) | byte[] |, | null (byte) |, ... } |
+   *
+   *
+   * This method does not permanently change the buffer position, limit, or mark.
+   *
+   * @return number of bytes written (1 if null, or 5 + size of String[] if not)
+   */
+  public static int writeNullableStringArray(ByteBuffer buffer, int offset, @Nullable String[] array, int maxSizeBytes)
+  {
+    // | null (byte) | array length (int) | array bytes |
+    if (array == null) {
+      return writeNull(buffer, offset);
+    } else {
+      int sizeBytes = 1 + Integer.BYTES;
+      buffer.put(offset, NullHandling.IS_NOT_NULL_BYTE);
+      buffer.putInt(offset + 1, array.length);
+      for (String element : array) {
+        if (element != null) {
+          final byte[] stringElementBytes = StringUtils.toUtf8(element);
+          checkMaxBytes(
+              ColumnType.STRING_ARRAY,
+              sizeBytes + 1 + Integer.BYTES + stringElementBytes.length,
+              maxSizeBytes
+          );
+          sizeBytes += writeNullableVariableBlob(buffer, offset + sizeBytes, stringElementBytes);
+        } else {
+          checkMaxBytes(
+              ColumnType.STRING_ARRAY,
+              sizeBytes + 1,
+              maxSizeBytes
+          );
+          sizeBytes += writeNull(buffer, offset + sizeBytes);
+        }
+      }
+      return sizeBytes;
+    }
+  }
+
+  /**
+   * Reads a nullable variably lengthed String[] value from a {@link ByteBuffer} at the supplied offset. If the null
+   * byte is set to {@link NullHandling#IS_NULL_BYTE}, this method will return null, else it will read the size of the
+   * array from the next 4 bytes and then read that many elements with {@link #readNullableVariableBlob} and decode them
+   * with {@link StringUtils#fromUtf8} to convert to string values.
+   *
+   * layout: | null (byte) | size (int) | {| null (byte) | size (int) | byte[] |, | null (byte) |, ... } |
+   *
+   * This method does not change the buffer position, limit, or mark.
+   */
+  @Nullable
+  public static String[] readNullableStringArray(ByteBuffer buffer, int offset)
+  {
+    // | null (byte) | array length (int) | array bytes |
+    if (isNullableNull(buffer, offset++)) {
+      return null;
+    }
+    final int stringArrayLength = buffer.getInt(offset);
+    offset += Integer.BYTES;
+    final String[] stringArray = new String[stringArrayLength];
+    for (int i = 0; i < stringArrayLength; i++) {
+      if (isNullableNull(buffer, offset)) {
+        stringArray[i] = null;
+      } else {
+        final byte[] stringElementBytes = readNullableVariableBlob(buffer, offset);
+        stringArray[i] = StringUtils.fromUtf8(stringElementBytes);
+        offset += Integer.BYTES + stringElementBytes.length;
+      }
+      offset++;
```

Comment:
Computing the new position is error-prone. Wouldn't it be easier to save the original position, compute the offset as the final position - the start position? Code would be simpler and we'd compute sizes once rather than twice.